### DISCUSSION TITLE
Behavior of ScenarioInitializer when enable concurrency

### DISCUSSION THREAD

**Discussion by nvcnvn** [cucumber/godog#452] [General]
```
func InitializeScenario(sc *godog.ScenarioContext) {
	sc.Before(func(ctx context.Context, sc *godog.Scenario) (context.Context, error) {
		Godogs = 0 // clean the state before every scenario

		return ctx, nil
	})

	sc.Step(`^there are (\d+) godogs$`, thereAreGodogs)
	sc.Step(`^I eat (\d+)$`, iEat)
	sc.Step(`^there should be (\d+) remaining$`, thereShouldBeRemaining)
}
```
Will this function called repeatedly for every goroutine? I add some dummy logs and it seem yes.
I think this hurting performance too much when you have a lot of step (we have 1000 steps).
Should it be on Suite level instead?

**Comment by vearutop** [comment]
I think it does not matter if the concurrency is enabled or not, scenario initializer runs for every scenario. 

This is an interesting idea, in many cases hooks and steps should be same for all scenario and indeed reusing them from suite can be a nice optimization.

Let me try a quick prototype.

**Comment by vearutop** [comment]
Please try `go get -u github.com/cucumber/godog@suite-level-hooks` in your project.
And then change your init from something like
```go
		ScenarioInitializer: func(s *godog.ScenarioContext) {
			local.RegisterSteps(s)
		},
```
to something like
```go
		TestSuiteInitializer: func(suiteContext *godog.TestSuiteContext) {
			local.RegisterSteps(suiteContext.ScenarioContext())
		},
```

Please share how much time does it take to run tests for both options.

**Reply by nvcnvn** [reply]
Oh it take the same amount of time.
Note that we try to do some minor optimization already, in our RegisterSteps helper we try to parse the regex Once and reuse.
```
func RegisterSteps(ctx *godog.ScenarioContext)
	steps := make(map[string]interface{}, 10000) // our 10k steps
	buildRegexpMapOnce.Do(func() { regexpMap = helper.BuildRegexpMap(steps) })
	for k, v := range steps {
		ctx.Step(regexpMap[k], v)
	}
}

func BuildRegexpMap(steps map[string]interface{}) map[string]*regexp.Regexp {
	m := make(map[string]*regexp.Regexp, len(steps))
	for stepDef, stepFunc := range steps {
		m[stepDef] = regexp.MustCompile(stepDef)
	}

	return m
}

```

So I guess the reflection didn't cost that much.
Note that our test case have a lot of network block so I guess the other can still benefit from this.

**Reply by vearutop** [reply]
Ok, I see. Yes, it is likely that `godog` overhead is very small comparing to test actions.

**Reply by nvcnvn** [reply]
But will you consider to keep this sir? I guess if we don't have the regex complied and cached it will take alot more CPU.

**Reply by vearutop** [reply]
Yes, I think this change still makes sense as a way to organize layering of global and per-scenario hooks/steps, I'll create a pull request.

**Reply by vearutop** [reply]
Though, currently scenario initializer lacks any information (like feature or scenario name) to decide whether a step should be registered or not, so it is always working in "global" mode with regards to steps/hooks.

**Reply by vearutop** [reply]
@nvcnvn please try [`v0.12.4`](https://github.com/cucumber/godog/releases/tag/v0.12.4).

**Reply by nvcnvn** [reply]
Thank you sir!
